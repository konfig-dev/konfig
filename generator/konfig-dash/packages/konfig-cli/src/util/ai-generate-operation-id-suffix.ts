import { Operation, OperationObject } from 'konfig-lib'
import OpenAI from 'openai'

export async function aiGenerateOperationIdSuffix({
  prefix,
  path,
  method,
  operation,
  operations,
}: {
  prefix: string
  path: string
  method: string
  operation: Operation
  operations: OperationObject[]
}): Promise<string> {
  const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] })

  const existingIds = operations.map((o) => o.operation.operationId)
  const existingIdsStr = existingIds.join(', ').slice(0, -2)
  const systemPrompt = `You are a generator of operation IDs for an OpenAPI specification. You generate operation IDs in json format: "{ prefix: ... , suffix: ...}"`
  const prompt = generatePrompt({
    prefix,
    path,
    method,
    operation,
    existingIdsStr,
  })

  const response = (
    await openai.chat.completions.create({
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      model: 'gpt-3.5-turbo-1106',
    })
  ).choices[0].message.content

  if (response === undefined || response === null || response === '')
    throw Error('OpenAI returned an empty response')

  const responseJson = JSON.parse(response)
  if (responseJson === undefined || responseJson === null)
    throw Error('OpenAI returned an invalid response: it is not a valid json')
  if (responseJson.suffix === undefined || responseJson.suffix === null)
    throw Error('OpenAI returned an invalid json: it does not contain a suffix')

  // Post-process to ensure that rules are followed
  let suffix = responseJson.suffix.replace('_', '')
  suffix = suffix.charAt(0).toLowerCase() + suffix.slice(1)
  suffix = suffix.replace('retrieve', 'get')
  suffix = suffix.replace('fetch', 'get')
  console.log(`AI-generated operation id: ${prefix}${suffix}`)
  return suffix
}

function generatePrompt({
  prefix,
  path,
  method,
  operation,
  existingIdsStr,
}: {
  prefix: string
  path: string
  method: string
  operation: Operation
  existingIdsStr: string
}): string {
  return `Generate an operation ID for this operation in my OpenAPI specification:
Path: ${path}
Method: ${method}
Summary: ${operation.summary}
Description: ${operation.description}

The operation ID is in the format "prefix_suffix". The prefix is: ${prefix}. Generate the suffix and respond in json format.
The operation ID MUST follow ALL of these rules:

1. The suffix must be at least 3 characters long.
2. The suffix must be in camelCase and CANNOT include any underscores.
3. The suffix must not redundantly re-use words from the prefix. (For example: Store_get is preferred over Store_getStore.)
4. If the suffix contains both a verb and a noun, the noun should generally follow the verb. (For example, Store_placeOrder is better than Store_orderPlace).
5. Prefer simple verbs like "get", "create", "update", "delete", and "list" when possible.

Existing operation ids for this OpenAPI specification are as follows: ${existingIdsStr}.
The operation ID MUST be unique within this list of existing operation IDs. You can also use them as a reference to generate this operation id.`
}
