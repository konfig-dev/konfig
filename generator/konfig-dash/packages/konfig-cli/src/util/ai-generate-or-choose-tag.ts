import { Operation, TagObject, OperationObject } from 'konfig-lib'
import OpenAI from 'openai'

export async function aiGenerateOrChooseTag(
  operation: Operation,
  path: string,
  method: string,
  tags: TagObject[]
): Promise<string> {
  const openai = new OpenAI({ apiKey: process.env['OPENAI_API_KEY'] })
  const tagNames = tags.map((t) => t.name)

  // First, try and choose a tag programmatically without using OpenAI
  const tag = programaticallyChooseTag(operation, path, tagNames)
  if (tag !== undefined) return tag
  const systemPrompt = `You are a generator of tags for an OpenAPI specification. You generate tags in json format: "{ tag: ... }"`
  const prompt = generatePrompt(path, method, operation, tagNames)
  const response = (
    await openai.chat.completions.create({
      response_format: { type: 'json_object' },
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: prompt },
      ],
      model: 'gpt-3.5-turbo-1106',
    })
  ).choices[0].message.content

  if (response === undefined || response === null || response === '')
    throw Error('OpenAI returned an empty response')

  const responseJson = JSON.parse(response)
  if (responseJson === undefined || responseJson === null)
    throw Error('OpenAI returned an invalid response: it is not a valid json')
  if (responseJson.tag === undefined || responseJson.tag === null)
    throw Error('OpenAI returned an invalid json: it does not contain a tag')

  tags.push({ name: responseJson.tag })
  return responseJson.tag
}

function programaticallyChooseTag(
  operation: Operation,
  path: string,
  tagNames: string[]
): string | undefined {
  for (const tag of tagNames) {
    if (path.toLowerCase().includes(tag.toLowerCase())) return tag
    if (
      operation.summary &&
      operation.summary.toLowerCase().includes(tag.toLowerCase())
    )
      return tag
  }
  return undefined
}

function generatePrompt(
  path: string,
  method: string,
  operation: Operation,
  tagNames: string[]
): string {
  const tagNamesStr = tagNames.join(', ').slice(0, -2)
  return `The following operation from my OpenAPI specification is missing a tag:
Path: ${path}
Method: ${method}
Summary: ${operation.summary}

Based on this information, choose the best tag from this list of existing tags on the OpenApi Specification: ${tagNamesStr}.
IF AND ONLY IF none of these existing tags are suitable, then generate a new tag. The new tag MUST be a single word, MUST be a noun, and MUST be singular. It MUST NOT be a combination of multiple words.

In general, a good tag name is related to the type of data being operated on (such as "Student"), or the type of operation being performed (such as "Authentication").

Respond with just the tag in json format: "{ tag: ... }" with no other information.`
}
