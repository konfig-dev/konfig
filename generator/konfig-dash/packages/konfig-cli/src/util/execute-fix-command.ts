import * as fs from 'fs-extra'
import {
  isJsonString,
  OBJECT_WITH_NO_PROPERTIES_RULE_NAME,
  parseSpec,
  filterSpecPaths,
  POTENTIAL_INCORRECT_DATA_TYPE_RULE_NAME,
} from 'konfig-lib'
import boxen from 'boxen'
import { Progress } from './fix-progress'
import { oasYamlDump } from './oas-yaml-dump'
import { fixOas } from './fix-oas'
import path from 'path'
import { oasJsonDump } from './oas-json-dump'
import { parseKonfigYaml } from './parse-konfig-yaml'

interface FixOptions {
  format?: boolean
  specInputPath?: string
  spec?: string
  auto?: boolean
  konfigDir?: string
  skipMissingResponseDescriptionFix?: boolean
  skipListUsageSecurity?: boolean
  alwaysYes?: boolean
  ci?: boolean
}

export async function executeFixCommand(options: FixOptions): Promise<void> {
  const flags = {
    ...options,
    auto: options.auto ?? true,
    alwaysYes: !!(options.alwaysYes == null || options.alwaysYes || options.ci), // ci mode should always confirm with yes
    ci: options.ci ?? false,
  }

  const { parsedKonfigYaml } = parseKonfigYaml({
    configDir: flags.konfigDir ?? process.cwd(),
  })

  // We only want to run fix in CI if the konfig.yaml has a specInputPath
  if (flags.ci && parsedKonfigYaml.specInputPath === undefined) {
    console.log(`No specInputPath found in konfig.yaml, skipping fix`)
    return
  }

  if (flags.spec === undefined) {
    flags.spec = parsedKonfigYaml.specPath
    flags.specInputPath = parsedKonfigYaml.specInputPath ?? flags.spec
  }

  if (flags.spec === undefined)
    throw Error(
      `Either specify path to OAS with -s flag or assign "specPath" field in "konfig.yaml"`
    )

  if (flags.specInputPath === undefined) throw Error(`This shouldn't happen`)

  const prependKonfigDir = ({ specPath }: { specPath: string }) => {
    return flags.konfigDir !== undefined
      ? path.join(flags.konfigDir, specPath)
      : specPath
  }

  const specInputPath = prependKonfigDir({ specPath: flags.specInputPath })
  const specOutputPath = prependKonfigDir({ specPath: flags.spec })
  const rawSpec = fs.readFileSync(specInputPath, 'utf-8')
  let spec = await parseSpec(rawSpec)

  // if konfig yaml filters any paths from the spec, remove them now
  if (parsedKonfigYaml.filterPaths) {
    filterSpecPaths({ spec: spec.spec, filter: parsedKonfigYaml.filterPaths })
  }

  if (flags.format) {
    fs.writeFileSync(flags.spec, oasYamlDump(spec))
    return
  }

  const progress: Progress = Progress.getSaved({
    konfigDir: flags.konfigDir ?? process.cwd(),
    progressYamlPathOverride: parsedKonfigYaml.progressYamlPath,
  })

  const {
    issuesFixed,
    infoDescriptionFixed,
    numberOfMissingResponseDescriptionsAdded,
    numberOfExamplesFixed,
    numberOfUpdatedOperationIds,
    numberOfMissingTags,
    numberOfTrailingSlashesFixed,
    numberOfMissing200ResponsesAdded,
    numberOfParametersConvertedToSecurityRequirements,
    numberOfEmptyRequestBodiesRemoved,
    numberOfListUsagesOfSecurityFixed,
    numberOfImproperlyNamedTags,
    numberOfRedundantSecurityAndParametersFixed,
    numberOfObjectsWithNoPropertiesFixed,
    numberOfSchemasDefined,
    numberOfParametersRemovedForNewSecurityScheme,
    numberOfUnstructuredRequestBodiesFixed,
    numberOfIgnorePotentialIncorrectTypeAdded,
    numberOfCustomModifications,
    numberOfIgnoreObjectsWithNoPropertiesAdded,
    numberOfEmptyResponseBodySchemasFixed,
    numberOfDuplicateTagNamesFixed,
    numberOfInvalidServerUrlsFixed,
    numberOfRedundantSecuritySchemesRemoved,
    numberOfExampleAndExamplesFixed,
    numberOfDisallowedHeaderNamesRemoved,
    numberOfNewTagNames,
    numberOfAnyOfTypeNullUsagesReverted,
    numberOfOas31UsagesFixed,
    numberOfExamplesUsageRemoved,
  } = await fixOas({
    spec,
    progress,
    konfigYaml: parsedKonfigYaml,
    alwaysYes: flags.alwaysYes,
    auto: flags.auto,
    ci: flags.ci,
    skipMissingResponseDescription: flags.skipMissingResponseDescriptionFix,
    skipFixListUsageSecurity: flags.skipListUsageSecurity,
  })

  fs.writeFileSync(
    specOutputPath,
    isJsonString(rawSpec) ? oasJsonDump(spec.spec) : oasYamlDump(spec.spec)
  )

  console.log(
    boxen(
      `Updated ${numberOfUpdatedOperationIds} operation IDs
Made ${numberOfCustomModifications} custom modifications
Renamed ${numberOfImproperlyNamedTags} improperly named tags
Removed ${numberOfDisallowedHeaderNamesRemoved} disallowed headers removed
Added ${numberOfIgnoreObjectsWithNoPropertiesAdded} ${OBJECT_WITH_NO_PROPERTIES_RULE_NAME} ignore rule
Added ${numberOfIgnorePotentialIncorrectTypeAdded} ${POTENTIAL_INCORRECT_DATA_TYPE_RULE_NAME} ignore rule
Added ${infoDescriptionFixed} missing info description
Added ${numberOfMissingResponseDescriptionsAdded} missing response descriptions
Added ${numberOfMissingTags} missing tags
Added ${numberOfEmptyResponseBodySchemasFixed} empty response body schemas
Added ${numberOfSchemasDefined} named schemas
Added ${numberOfMissing200ResponsesAdded} missing 2xx responses
Added ${numberOfNewTagNames} new tag names
Fixed ${numberOfDuplicateTagNamesFixed} duplicate tag names fixed
Fixed ${numberOfInvalidServerUrlsFixed} invalid server URLs fixed
Fixed ${numberOfUnstructuredRequestBodiesFixed} unstructured request bodies
Fixed ${numberOfObjectsWithNoPropertiesFixed} objects with no properties
Fixed ${numberOfExampleAndExamplesFixed} redundant "example" & "examples" fields
Fixed ${numberOfExamplesFixed} examples with invalid schemas
Fixed ${numberOfParametersConvertedToSecurityRequirements} parameters that should be security requirements
Fixed ${numberOfListUsagesOfSecurityFixed} list usages of security
Removed ${numberOfTrailingSlashesFixed} trailing slashes
Removed ${numberOfRedundantSecurityAndParametersFixed} redundant security requirement and parameters
Removed ${numberOfRedundantSecuritySchemesRemoved} redundant security schemes
Removed ${numberOfParametersRemovedForNewSecurityScheme} parameters replace with security requirement
Removed ${numberOfEmptyRequestBodiesRemoved} empty request bodies
Fixed ${numberOfAnyOfTypeNullUsagesReverted} anyOf type null usages
Fixed ${numberOfOas31UsagesFixed} OAS 3.1.x usages
Fixed ${numberOfExamplesUsageRemoved} "examples" usage`,
      {
        title: `Fixed ${issuesFixed} Issues`,
        borderColor: issuesFixed === 0 ? 'gray' : 'green',
        padding: 1,
      }
    )
  )
}
