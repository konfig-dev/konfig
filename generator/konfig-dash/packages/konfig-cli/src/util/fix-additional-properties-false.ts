import { Spec, recurseObject } from 'konfig-lib'

/**
 * "additionalProperties: false" causes issue when mocking an API.  The
 * following type of errors will appear for no reason: "âœ–  error     Request
 * body must NOT have additional properties; found 'userId'" but userId is
 * actually a property in the schema so the error is just wrong.  To get around
 * this I added this rule to remove "additionalProperties: false" from the
 * schema.
 */
export function fixAdditionalPropertiesFalse({ spec }: { spec: Spec }): number {
  let numberOfAdditionalPropertiesFixed = 0
  recurseObject(spec.spec, ({ value: schema }) => {
    if (typeof schema !== 'object') return
    if (schema === null) return
    if (schema.type !== 'object') return
    if (schema.additionalProperties === undefined) return
    if (schema.additionalProperties === true) return
    delete schema.additionalProperties
    numberOfAdditionalPropertiesFixed++
  })
  return numberOfAdditionalPropertiesFixed
}
