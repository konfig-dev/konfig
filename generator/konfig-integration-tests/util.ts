import { execa } from "execa";
import * as path from "path";
import yaml from "js-yaml";
import { expect } from "vitest";
import * as fs from "fs";
import {
  KONFIG_YAML_NAME,
  KonfigYaml,
  KonfigYamlAdditionalGeneratorConfig,
  KonfigYamlGeneratorConfig,
  KonfigYamlType,
} from "konfig-lib";
import waitOn from "wait-on";

const USE_RUN =
  process.env.GITHUB_ACTIONS === "true" ||
  (process.env.RUN && process.env.RUN.trim() !== "");

console.log("USE_RUN", USE_RUN);

// relative path from this file to: ""../konfig-dash/packages/konfig-cli/bin/dev""
// use nodejs __dirname to get the current directory of this file
const KONFIG_CLI_PATH = path.join(
  __dirname,
  `../konfig-dash/packages/konfig-cli/bin/${USE_RUN ? "run" : "dev"}`
);

/**
 * parse 'python-pydantic-empty-response' from 'tests/python-pydantic-empty-response.test.ts > python-pydantic-empty-response'
 */
function getCurrentTestName(): string {
  const currentTestName = expect.getState().currentTestName;
  const parsed = currentTestName?.split(" > ")[1];
  if (!parsed) {
    throw new Error(`Unable to parse current test name: ${currentTestName}`);
  }
  return parsed;
}

export async function e2e(
  mockServerPort: number,
  customAssertions?: () => void
) {
  if (!getCurrentTestName()) {
    throw new Error("Unable to get current test name");
  }
  const sdkDir = path.join(__dirname, "sdks", getCurrentTestName());

  const apiUrl = process.env.KONFIG_API_URL ?? "http://127.0.0.1:8911";

  // wait until KONFIG_API_URL/healthz is available
  await waitOn({
    resources: [`${apiUrl}/healthz`],
    timeout: 100000,
  });

  const env = {
    // This ensure oclif doesn't bug out
    // See: https://github.com/oclif/oclif/issues/1161#issuecomment-1661372245
    NODE_ENV: "",
  };

  // run "konfig generate -d -I" inside the path
  await execa(KONFIG_CLI_PATH, ["generate", "-d", "-I"], {
    cwd: sdkDir,
    stdio: "inherit",
    env,
  });

  // run "konfig test" inside the path
  await execa(KONFIG_CLI_PATH, ["test", "-p", mockServerPort.toString()], {
    cwd: sdkDir,
    stdio: "inherit",
    env,
  });

  // validate top-level README.md
  const konfigYamlPath = path.join(sdkDir, KONFIG_YAML_NAME);
  if (!fs.existsSync(konfigYamlPath))
    throw new Error(
      `Make sure you are in same directory as '${KONFIG_YAML_NAME}'`
    );
  const konfigYamlRaw = fs.readFileSync(konfigYamlPath, "utf-8");
  const loadedKonfigYaml: KonfigYamlType = yaml.load(
    konfigYamlRaw
  ) as KonfigYamlType;
  const parsedKonfigYaml = KonfigYaml.parse(loadedKonfigYaml);

  const generators: (
    | KonfigYamlGeneratorConfig
    | KonfigYamlAdditionalGeneratorConfig
  )[] = [
    ...Object.values(parsedKonfigYaml.generators),
    ...(parsedKonfigYaml.additionalGenerators !== undefined
      ? Object.values(parsedKonfigYaml.additionalGenerators)
      : []),
  ];
  for (const generator of generators) {
    expect(
      normalizeDocumentation(
        fs.readFileSync(
          path.join(sdkDir, generator.outputDirectory, "README.md"),
          "utf-8"
        )
      )
    ).toMatchSnapshot();
  }

  if (customAssertions) {
    customAssertions();
  }
}

// Removes the [GitHub last commit] line from the README
function normalizeDocumentation(readme: string) {
  // matches [GitHub last commit] to the end of the line
  const pattern = /\[!\[GitHub last commit\].*$\n?/gm;
  return readme.replace(pattern, "");
}
