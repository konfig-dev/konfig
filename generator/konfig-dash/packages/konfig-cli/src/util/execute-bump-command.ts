import { CliUx } from '@oclif/core'
import { getPackageVersion, setPackageVersion, KonfigYamlType } from 'konfig-lib'
import { generateKonfigYamlString } from '../util/generate-konfig-yaml-string'
import { incrementVersionWithPreservedTag } from '../util/increment-version-with-preserved-tag'
import semver from 'semver'
import * as fs from 'fs-extra'

export async function executeBumpCommand({
  filter,
  releaseType,
  loadedKonfigYaml,
  konfigYamlPath,
}: {
  filter: string[] | null
  releaseType: semver.ReleaseType
  loadedKonfigYaml: KonfigYamlType
  konfigYamlPath: string
}): Promise<void> {

  if (filter !== null)
    CliUx.ux.info(`Performing ${releaseType} bump of ${filter.join(', ')}`)
  else CliUx.ux.info(`Performing "${releaseType}" bump`)

  for (const [generatorName, generatorConfig] of [
    ...Object.entries(loadedKonfigYaml.generators),
    ...Object.entries(loadedKonfigYaml.additionalGenerators ?? []),
  ]) {
    if (filter !== null && !filter.includes(generatorName)) continue
    if ('disabled' in generatorConfig && generatorConfig.disabled) continue
    const currentVersion = getPackageVersion({ generatorConfig })
    const newVersion = incrementVersionWithPreservedTag(currentVersion, releaseType)
    CliUx.ux.info(`Bumping ${generatorName} SDK from ${currentVersion} --> ${newVersion}`)
    if (newVersion === null)
      CliUx.ux.error(`Could not compute new version for ${generatorName} SDK currently at ${currentVersion}`)
    setPackageVersion({ generatorConfig, version: newVersion })
  }
  CliUx.ux.action.start('Writing new version to konfig.yaml')
  fs.writeFileSync(konfigYamlPath, generateKonfigYamlString(loadedKonfigYaml))
  CliUx.ux.action.stop()
}
