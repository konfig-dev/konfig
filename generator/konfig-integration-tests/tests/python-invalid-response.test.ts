import { e2e } from "../util";
import { test } from "vitest";

test("python-invalid-response", async () => {
  await e2e(4027, {
    customServer: {
      routes: [
        {
          path: "/invalid-scalar",
          method: "get",
          response: {
            str: 2,
          },
        },
        {
          path: "/invalid-array",
          method: "get",
          response: {
            array: {
              hello: "world",
            },
          },
        },
        {
          path: "/invalid-object",
          method: "get",
          response: {
            object: [
              {
                hello: "world",
              },
            ],
          },
        },
      ],
    },
  });
});
