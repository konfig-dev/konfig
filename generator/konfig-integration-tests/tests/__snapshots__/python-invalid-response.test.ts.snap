// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`python-invalid-response 1`] = `
"# python-invalid-response<a id=\\"python-invalid-response\\"></a>

A simple API based for testing python-invalid-response.


[![PyPI](https://img.shields.io/badge/PyPI-v1.0.0-blue)](https://pypi.org/project/python-invalid-response/1.0.0)
[![README.md](https://img.shields.io/badge/README-Click%20Here-green)](https://github.com/konfig-dev/konfig/tree/main/python#readme)
[![More Info](https://img.shields.io/badge/More%20Info-Click%20Here-orange)](http://example.com/support)

## Table of Contents<a id=\\"table-of-contents\\"></a>

<!-- toc -->

- [Requirements](#requirements)
- [Installation](#installation)
- [Getting Started](#getting-started)
- [Async](#async)
- [Raw HTTP Response](#raw-http-response)
- [Reference](#reference)
  * [\`pythoninvalidresponseclient.test.dict_instead_of_list_or_scalar\`](#pythoninvalidresponseclienttestdict_instead_of_list_or_scalar)
  * [\`pythoninvalidresponseclient.test.invalid_array\`](#pythoninvalidresponseclienttestinvalid_array)
  * [\`pythoninvalidresponseclient.test.invalid_object\`](#pythoninvalidresponseclienttestinvalid_object)
  * [\`pythoninvalidresponseclient.test.invalid_scalar\`](#pythoninvalidresponseclienttestinvalid_scalar)
  * [\`pythoninvalidresponseclient.test.list_instead_of_scalar\`](#pythoninvalidresponseclienttestlist_instead_of_scalar)
  * [\`pythoninvalidresponseclient.test.object_instead_of_scalar\`](#pythoninvalidresponseclienttestobject_instead_of_scalar)
  * [\`pythoninvalidresponseclient.test.only_one_property_is_invalid\`](#pythoninvalidresponseclienttestonly_one_property_is_invalid)

<!-- tocstop -->

## Requirements<a id=\\"requirements\\"></a>

Python >=3.7

## Installation<a id=\\"installation\\"></a>

\`\`\`sh
pip install python-invalid-response==1.0.0
\`\`\`

## Getting Started<a id=\\"getting-started\\"></a>

\`\`\`python
from pprint import pprint
from python_invalid_response import PythonInvalidResponseClient, ApiException

pythoninvalidresponseclient = PythonInvalidResponseClient(
    api_key_auth=\\"YOUR_API_KEY\\",
)

try:
    dict_instead_of_list_or_scalar_response = (
        pythoninvalidresponseclient.test.dict_instead_of_list_or_scalar()
    )
    print(dict_instead_of_list_or_scalar_response)
except ApiException as e:
    print(\\"Exception when calling TestApi.dict_instead_of_list_or_scalar: %s\\\\n\\" % e)
    pprint(e.body)
    pprint(e.headers)
    pprint(e.status)
    pprint(e.reason)
    pprint(e.round_trip_time)
\`\`\`

## Async<a id=\\"async\\"></a>

\`async\` support is available by prepending \`a\` to any method.

\`\`\`python
import asyncio
from pprint import pprint
from python_invalid_response import PythonInvalidResponseClient, ApiException

pythoninvalidresponseclient = PythonInvalidResponseClient(
    api_key_auth=\\"YOUR_API_KEY\\",
)


async def main():
    try:
        dict_instead_of_list_or_scalar_response = (
            await pythoninvalidresponseclient.test.adict_instead_of_list_or_scalar()
        )
        print(dict_instead_of_list_or_scalar_response)
    except ApiException as e:
        print(\\"Exception when calling TestApi.dict_instead_of_list_or_scalar: %s\\\\n\\" % e)
        pprint(e.body)
        pprint(e.headers)
        pprint(e.status)
        pprint(e.reason)
        pprint(e.round_trip_time)


asyncio.run(main())
\`\`\`

## Raw HTTP Response<a id=\\"raw-http-response\\"></a>

To access raw HTTP response values, use the \`.raw\` namespace.

\`\`\`python
from pprint import pprint
from python_invalid_response import PythonInvalidResponseClient, ApiException

pythoninvalidresponseclient = PythonInvalidResponseClient(
    api_key_auth=\\"YOUR_API_KEY\\",
)

try:
    dict_instead_of_list_or_scalar_response = (
        pythoninvalidresponseclient.test.raw.dict_instead_of_list_or_scalar()
    )
    pprint(dict_instead_of_list_or_scalar_response.body)
    pprint(dict_instead_of_list_or_scalar_response.headers)
    pprint(dict_instead_of_list_or_scalar_response.status)
    pprint(dict_instead_of_list_or_scalar_response.round_trip_time)
except ApiException as e:
    print(\\"Exception when calling TestApi.dict_instead_of_list_or_scalar: %s\\\\n\\" % e)
    pprint(e.body)
    pprint(e.headers)
    pprint(e.status)
    pprint(e.reason)
    pprint(e.round_trip_time)
\`\`\`


## Reference<a id=\\"reference\\"></a>
### \`pythoninvalidresponseclient.test.dict_instead_of_list_or_scalar\`<a id=\\"pythoninvalidresponseclienttestdict_instead_of_list_or_scalar\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`python
dict_instead_of_list_or_scalar_response = (
    pythoninvalidresponseclient.test.dict_instead_of_list_or_scalar()
)
\`\`\`

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[\`DictInsteadOfListOrScalar\`](./python_invalid_response/pydantic/dict_instead_of_list_or_scalar.py)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/dict-instead-of-list-or-scalar\` \`get\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---

### \`pythoninvalidresponseclient.test.invalid_array\`<a id=\\"pythoninvalidresponseclienttestinvalid_array\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`python
invalid_array_response = pythoninvalidresponseclient.test.invalid_array()
\`\`\`

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[\`InvalidArray\`](./python_invalid_response/pydantic/invalid_array.py)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/invalid-array\` \`get\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---

### \`pythoninvalidresponseclient.test.invalid_object\`<a id=\\"pythoninvalidresponseclienttestinvalid_object\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`python
invalid_object_response = pythoninvalidresponseclient.test.invalid_object()
\`\`\`

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[\`InvalidObject\`](./python_invalid_response/pydantic/invalid_object.py)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/invalid-object\` \`get\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---

### \`pythoninvalidresponseclient.test.invalid_scalar\`<a id=\\"pythoninvalidresponseclienttestinvalid_scalar\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`python
invalid_scalar_response = pythoninvalidresponseclient.test.invalid_scalar()
\`\`\`

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[\`InvalidScalar\`](./python_invalid_response/pydantic/invalid_scalar.py)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/invalid-scalar\` \`get\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---

### \`pythoninvalidresponseclient.test.list_instead_of_scalar\`<a id=\\"pythoninvalidresponseclienttestlist_instead_of_scalar\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`python
list_instead_of_scalar_response = (
    pythoninvalidresponseclient.test.list_instead_of_scalar()
)
\`\`\`

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/list-instead-of-scalar\` \`get\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---

### \`pythoninvalidresponseclient.test.object_instead_of_scalar\`<a id=\\"pythoninvalidresponseclienttestobject_instead_of_scalar\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`python
object_instead_of_scalar_response = (
    pythoninvalidresponseclient.test.object_instead_of_scalar()
)
\`\`\`

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/object-instead-of-scalar\` \`get\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---

### \`pythoninvalidresponseclient.test.only_one_property_is_invalid\`<a id=\\"pythoninvalidresponseclienttestonly_one_property_is_invalid\\"></a>



#### 🛠️ Usage<a id=\\"🛠️-usage\\"></a>

\`\`\`python
only_one_property_is_invalid_response = (
    pythoninvalidresponseclient.test.only_one_property_is_invalid()
)
\`\`\`

#### 🔄 Return<a id=\\"🔄-return\\"></a>

[\`OnlyOnePropertyIsInvalid\`](./python_invalid_response/pydantic/only_one_property_is_invalid.py)

#### 🌐 Endpoint<a id=\\"🌐-endpoint\\"></a>

\`/only-one-property-is-invalid\` \`get\`

[🔙 **Back to Table of Contents**](#table-of-contents)

---


## Author<a id=\\"author\\"></a>
This Python package is automatically generated by [Konfig](https://konfigthis.com)
"
`;
